### PR TITLE
Upgrade to jline3

### PR DESCRIPTION
**Pull request by Manno15**
There is still some work to be done to this pull request but I wanted to get this out there to get some feedback in some areas.

I still need to do more robust testing on the shell and go back to a few areas that I commented out. I will also refactor/optimize areas if possible.

### CODE REVIEW

**Comment by ctubbsii** [comment]
@Manno15 This is a great start. It looks like a beast of a problem, but it looks like you're headed in the right direction! The fact that so much of the API of JLine is subject to change makes me wonder if we want to just create our own internal abstraction... like a single class that wraps all the JLine stuff, and puts it all in one place, and provides more friendly methods to get to what we need. I'm thinking something along the lines of:

```java
  class SomeClassName {
    SomeClassName() {
      // set up JLine terminal and related objects here
      // no JLine types used outside this class
      // if we need anything from JLine, we add a convenience method below, so JLine is entirely accessed through this class
    }
    println(...) { ... }
    readline(...) { ... }
    saveHistory(...) { ... }
  }

  var console = new SomeClassName();
  console.println("something")
```

**Comment by Manno15** [comment]
Yeah, that sounds like a great idea @ctubbsii. I will look into to implementing something like that while I continue to fix some of the areas I was unsure on.

**Comment by Manno15** [comment]
This is getting close to being finished. Some issues I am facing are history file incompatibilities from previous versions (I believe jline3 requires timestamps in their history file) and hitting `'q'` while in the help command doesn't exit out of the command until you hit `enter` after the `'q'`.

**Comment by Manno15** [comment]
I believe this pull request is ready for review. There are a couple of areas where I unsure of the proper route to go on. One is the use of `DumbTerminal` in `ShellIT` and `ShellServerIT`. The other is [here](https://github.com/apache/accumulo/blob/61a1755547fc0f9fd6290eb4abe704b775657618/shell/src/main/java/org/apache/accumulo/shell/Shell.java#L563)

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java
Comment (DomGarguilo) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java:278]: Does anyone know why the print statements are like this as opposed to one/fewer prints with linebreaks (`\n`). It seems as though there might be a reason `println`s are used in this way, but it seems like it would be cleaner to consolidate this text into less prints if possible. I see that it was this way before this PR so its not quite related but I was just curious.

Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java:278]: Sometimes this is done for readability, sometimes to reuse a buffer for performance to avoid unnecessary string concatenation (as in StringBuilder append methods), and sometimes to defer to environment-specific implementations (like cross-platform newline differences). It could also just be more convenient for a developer to copy/paste as many `println` calls as they want, rather than edit in a specific number of `\n` in several different String literals.

I'm not sure why it's used this way in this particular case, but probably some combination of readability and developer convenience. But, it probably also makes sense in this case to defer to the terminal's environment-specific implementation, since the terminal itself is intended as an abstraction around the command-line environment.

Comment (DomGarguilo) [reply server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java:278]: Okay great. Thanks for the explanation.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (DomGarguilo) [inline shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:88]: Could the portion before `&&` be replaced with `more`?
```suggestion
        delete = more && (line.equalsIgnoreCase("y") || line.equalsIgnoreCase("yes"));
```

**Comment by DomGarguilo** [comment]
Looks good from what I can tell. Manually ran a lot of the commands and everything appears as expected.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (lbschanno) [inline shell/src/main/java/org/apache/accumulo/shell/Shell.java:620]: I suggest adding a log.warn or log.error here to capture the exception within the logs.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (dlmarion) [inline shell/src/main/java/org/apache/accumulo/shell/Shell.java:562]: Did you test CTRL-C without the lines below?

**Review by dlmarion — APPROVED** [review]
LGTM, I made a comment about testing some code that you thought could be removed.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:562]: I believe CTRL-C still worked without it but I will double check tomorrow to verify.

File: test/src/main/java/org/apache/accumulo/test/ShellServerIT.java
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:201]: I wonder if it would make more sense to pass in "terminal" instead of just the reader?

File: test/src/main/java/org/apache/accumulo/test/ShellIT.java
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/ShellIT.java:220]: If these tests pass without the newline, then that seems like it would indicate a breaking change to getsplits.

File: shell/src/test/java/org/apache/accumulo/shell/format/DeleterFormatterTest.java
Comment (ctubbsii) [inline shell/src/test/java/org/apache/accumulo/shell/format/DeleterFormatterTest.java:144]: What's this comment about?

File: shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java
Comment (ctubbsii) [inline shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java:74]: Is there a way to add the history directly to the terminal, and using its line reader, or do we have to create a new line reader separately from the terminal?

File: shell/src/test/java/org/apache/accumulo/shell/ShellConfigTest.java
Comment (ctubbsii) [inline shell/src/test/java/org/apache/accumulo/shell/ShellConfigTest.java:32]: Is it safe to use the impl here? Doing so seems fragile.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:85]: Which one? Why? Can it be fixed?

File: shell/src/main/java/org/apache/accumulo/shell/commands/HistoryCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/HistoryCommand.java:46]: It looks like Iterator was used here previously because `.entries()` returned a type of `Iterator`. However, it now looks like `.getHistory()` returns a type of `Iterable`, which gives an opportunity to write this differently, using Java streams or something, and to avoid the use of Guava's `Iterators.transform()` in favor of Java built-ins. It's not important, though. It's fine either way.

File: shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java:46]: Have you tested the bell to make sure it actually rings? (might depend on your local development environment).

File: shell/src/main/java/org/apache/accumulo/shell/ShellCompletor.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/ShellCompletor.java:50]: These warning suppressions might not make sense anymore, now that you've added the appropriate generics to the candidates list.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (ctubbsii) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:620]: I concur. We don't generally just print stack traces with `e.printStackTrace()`.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/Shell.java:598]: This println previously existed, so the user's bash shell or whatever, wouldn't print its prompt at the end of Accumulo's prompt, when existing by closing the input stream. I assume by removing this println, that no longer occurs? If it does, that should probably be put back.

**Review by ctubbsii — CHANGES_REQUESTED** [review]
Most of this looks great. I am concerned about the getsplits behavior and corresponding tests, though.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:620]: I will add.

File: shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java:46]: I have not but I will try on my end. This was how the devs on the mailing list said to do it. I never heard the bell on the old shell so I am not sure if I will hear this one.

**Comment by Manno15** [comment]
I am running into an issue getting these changes to run with zookeeper 3.6.2 on fluo-uno. I will investigate. Specific error below: 

<pre>
Exception in thread "shell" java.lang.UnsatisfiedLinkError: 'int org.fusesource.jansi.internal.CLibrary.ioctl(int, long, org.fusesource.jansi.internal.CLibrary$WinSize)'
	at org.fusesource.jansi.internal.CLibrary.ioctl(Native Method)
	at org.jline.terminal.impl.jansi.JansiNativePty.getSize(JansiNativePty.java:146)
	at org.jline.terminal.impl.AbstractPosixTerminal.getSize(AbstractPosixTerminal.java:60)
	at org.jline.terminal.Terminal.getBufferSize(Terminal.java:216)
	at org.jline.reader.impl.LineReaderImpl.doDisplay(LineReaderImpl.java:746)
	at org.jline.reader.impl.LineReaderImpl.<init>(LineReaderImpl.java:303)
	at org.jline.reader.LineReaderBuilder.build(LineReaderBuilder.java:115)
	at org.apache.accumulo.shell.Shell.config(Shell.java:275)
	at org.apache.accumulo.shell.Shell.execute(Shell.java:520)
	at org.apache.accumulo.start.Main.lambda$execKeyword$0(Main.java:126)
	at java.base/java.lang.Thread.run(Thread.java:834)

</pre>

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:562]: In my testing, both the main branch and these changes do have the same functionality for CTRL-C but they both might be wrong. When I  hit CTRL-C, it exits the shell which I do not think is what we are going for. I tested the jline3 demo and CTRL-C will exit back to the shell but not exit the shell. That is probably the functionality we want. The funny thing is, I am implementing what they did in their demo but for some reason, it does not seem to be working. So this will take some more investigating.

Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:562]: I did find that when I CTRL-C in the middle of the help command, it does work properly. As in it throws the interrupt exception but doesn't exit the process. It only exits if I CTRL-C when just sitting in the shell.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java]: I noticed a printExcetion function included in the shell class already so I used that for this. Let me know if that suffices

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java
Comment (Manno15) [reply server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Done in latest commit

File: test/src/main/java/org/apache/accumulo/test/ShellServerIT.java
Comment (Manno15) [reply test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:201]: I am unsure if it would change anything since the terminal gets attached to the reader. We could just pass in "terminal" and then build the reader inside `Shell.java` so keep things consistent there. Not sure what that would gain ultimately.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:85]: This is related to the other comment you mentioned in the `DeleterFormatterTest`. I will have to refresh my memory on it a bit.

File: shell/src/test/java/org/apache/accumulo/shell/format/DeleterFormatterTest.java
Comment (Manno15) [reply shell/src/test/java/org/apache/accumulo/shell/format/DeleterFormatterTest.java:144]: Related to your other comment below in `DeleterFormatter`.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:85]: the failure happens on line 86 where it tries to read the delete prompt.  When `DeleterFormatterTest.testNoConfirmation()` is ran, it hits that line but gets an end of file marker instead.

Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:85]: It is possible that Jline3 changed something behind the scenes. On the main branch, line returns null when trying to read the prompt. Currently, it reads the end of file. Not sure what the best approach to fix it yet but I am trying to get it to where it returns null.

File: shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java
Comment (ctubbsii) [reply shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java:46]: The bell should be the same one you hear in Linux if you got backspace repeatedly on an empty line. If you can't hear it by trying that, then it's probably configurable in your system sound settings to enable it.

File: test/src/main/java/org/apache/accumulo/test/ShellServerIT.java
Comment (ctubbsii) [reply test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:201]: The gain is just more intuitive API semantics (possibly) and maybe exposing a more stable jline API... a terminal being a high level object and line reader a lower level component.

Comment (Manno15) [reply test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:201]: I will look into it further once I get through the rest of the comments.

File: shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/commands/HiddenCommand.java:46]: Confirmed that the bell is there with these changes. Had to do a bit of settings adjustments in my terminal to make it audible.

File: test/src/main/java/org/apache/accumulo/test/ShellServerIT.java
Comment (Manno15) [reply test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:907]: Yep, that was my mistake. Apparently, in `PrintShell` I replaced a printLn() with just a print().

File: test/src/main/java/org/apache/accumulo/test/ShellIT.java
Comment (Manno15) [reply test/src/main/java/org/apache/accumulo/test/ShellIT.java:220]: Same issue as above. Printed instead of PrintLn

File: shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java
Comment (Manno15) [reply shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java:74]: The line reader is what contains the history and also contains the terminal object. Not the other way around. The line reader seems to be the higher-level object though I agree that it doesn't make the most sense. It is probably like this so it doesn't differentiate itself too much from `ConsoleReader` in jline2. 

The reader object in terminal is a bit different than a LineReader and isn't as friendly to use.

File: shell/src/test/java/org/apache/accumulo/shell/ShellConfigTest.java
Comment (Manno15) [reply shell/src/test/java/org/apache/accumulo/shell/ShellConfigTest.java:32]: I forgot to update this test to use a dumb terminal like the other tests.

File: shell/src/main/java/org/apache/accumulo/shell/commands/HistoryCommand.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/commands/HistoryCommand.java:46]: This is something I can look for additional work after this PR. I did briefly look at trying to implement it and noticed a decent amount of code to change at where we interpret the results and print to shell.

File: shell/src/main/java/org/apache/accumulo/shell/ShellCompletor.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/ShellCompletor.java]: Removed in the latest commit

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:600]: I don't believe this still occurs but I just noticed that CTRL-D will throw an `EndOfFileException` so I'll add a catch for that exception.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java]: This is because of the block below. Not sure if this is a feature that we want or not.

```java
catch (UserInterruptException uie) {
        // User Cancelled (Ctrl+C)
        reader.println();
        writer.println();

        String partialLine = uie.getPartialLine();
        if (partialLine == null || "".equals(uie.getPartialLine().trim())) {
          // No content, actually exit
          return exitCode;
        }
```

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java]: I added a potential fix for this in my latest commit. It is probably good enough without completely rewriting `DeleterFormatter`

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (ctubbsii) [inline shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:93]: I think this suggestion was made previously, but may have been overlooked.
```suggestion
      delete = more && (line.equalsIgnoreCase("y") || line.equalsIgnoreCase("yes"));
```

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (ctubbsii) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:600]: > I don't believe this still occurs but I will add it back because I don't have a great way to test and a println doesn't hurt.

Manually/visual testing for this is sufficient.

> I also just noticed that CTRL-D will throw an `EndOfFileException` if I do it on a blank line so I'll add a catch for that exception. This might make that if statement redundant if that is the only situation input could be null.

Hmm, yeah, if it's causing that to explicitly be thrown, then it might be redundant. Handling it by catching the exception is much more clear. As long as the output doesn't get jumbled, and we get a new line when Ctrl-D is entered, I'm not too concerned about how the handling is implemented, as long as the behavior works.

File: shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java
Comment (ctubbsii) [reply shell/src/test/java/org/apache/accumulo/shell/commands/HistoryCommandTest.java:74]: Makes sense! Thanks for the explanation.

File: test/src/main/java/org/apache/accumulo/test/ShellServerIT.java
Comment (ctubbsii) [reply test/src/main/java/org/apache/accumulo/test/ShellServerIT.java:201]: I think your other comment explained that LineReader was more the primary object to interact with, not the terminal. If that's the case, then my suggestion here probably is fine to ignore.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java:93]: You are correct. I will add that in.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:600]: I will do more testing to confirm then push out my findings, CTRL-D for sure is hit by the exception and everything looks good there. I just need to make sure there is no other case input will equal null and to see what happens if something like that does happen.

File: shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/format/DeleterFormatter.java]: I will add this. Sorry for the overlook.

File: shell/src/main/java/org/apache/accumulo/shell/Shell.java
Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:600]: Another question I have, is if we want to exit on CTRL-D or for CTRL-C for that matter. I am in favor of not exiting on either of those two things but the current implementation is to exit the shell.

Comment (Manno15) [reply shell/src/main/java/org/apache/accumulo/shell/Shell.java:600]: From my testing, there is no reasonable way for a user to make input null beyond hitting CTRL-D, which sends a EOF exception. I interpret that as the user wanting to exit (which is standard for BASH but we can change that if we want). My latest commit reflects that idea.

**Comment by Manno15** [comment]
@ctubbsii  I believe I hit on everyones questions and suggestions. Do you have anything further to add or any concerns?

**Comment by ctubbsii** [comment]
> @ctubbsii I believe I hit on everyones questions and suggestions. Do you have anything further to add or any concerns?

I can take another look today and re-run the ITs.

**Comment by ctubbsii** [comment]
> > @ctubbsii I believe I hit on everyones questions and suggestions. Do you have anything further to add or any concerns?
> 
> I can take another look today and re-run the ITs.

I re-ran all the ITs, and they all pass, except for a few known flaky ones.

**Review by ctubbsii — CHANGES_REQUESTED** [review]
I think this is probably mostly okay to merge now. However, I haven't done any manual testing... only code inspection and running tests. So, we should be careful to observe any behavioral changes to the shell as we continue development/testing on 2.1.

There is one more thing that does need to be done before this can be merged, though:
* the LICENSE file in the assembly tarball needs to be updated (assemble/src/main/resources/LICENSE) based on the upstream license information (there are only slight differences with https://github.com/jline/jline3/blob/jline-parent-3.17.1/LICENSE.txt

By the way, you may wish to consider updating to 3.19.0, in case some things have improved since you started working on this.

**Comment by Manno15** [comment]
> I think this is probably mostly okay to merge now. However, I haven't done any manual testing... only code inspection and running tests. So, we should be careful to observe any behavioral changes to the shell as we continue development/testing on 2.1.

Understandable. Fortunately, most of the commands themselves have stayed the same. The things that have changed the most is tab completion and how the history file is set. Those I have tested thoroughly. The tests have also changed quite a bit and possible improvements could be made to those in future versions of Jline. 

Another possible concern is how the terminal and reader will be set up on different machines. We let Jline determine the best terminal type and use the default implementations for other various features that maybe we would want to add later on.  This isn't different from what we did in Jline2 so it isn't a big concern but I have already noticed several changes in functionality between `ConsoleReader` and the `terminal`/`lineReader` combo that is used now.  

> There is one more thing that does need to be done before this can be merged, though:
  the LICENSE file in the assembly tarball needs to be updated (assemble/src/main/resources/LICENSE) based on the upstream license information (there are only slight differences with https://github.com/jline/jline3/blob/jline-parent-3.17.1/LICENSE.txt
By the way, you may wish to consider updating to 3.19.0, in case some things have improved since you started working on this

I will add these Monday. I am curious to see what they changed in 3.19.

**Comment by Manno15** [comment]
I updated the license and upgraded to 3.19.0. There wasn't anything noticeable to the upgrade that would benefit us so no further changes from it. I will merge this today unless there are any objections.

**Review by ctubbsii — APPROVED** [review]

**Pull request merged**
